### PR TITLE
Benchmarks: Add benchmark - add source code of cudnn function micro benchmark

### PR DESCRIPTION
**Pull request by yukirora**
**Description**
add source code of cudnn function microbenchmark

### CODE REVIEW

File: superbench/benchmarks/micro_benchmarks/cudnn_test/para_info.json [hunk 1]
```diff
@@ -0,0 +1,920 @@
+[
```
Comment (cp5555) [inline superbench/benchmarks/micro_benchmarks/cudnn_test/para_info.json:1]: Remove the json first.

File: superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_helper.cpp [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        // Make sure we call CUDA Device Reset before exiting
+        throw std::runtime_error(safe_call_ss.str());
+    }
+}
+/**
+ * @brief Cuda context init
+ */
+void cuda_init(cudnnHandle_t *cudnn_handle) {
+    CUDA_SAFE_CALL(cudaDeviceReset());
+    CUDA_SAFE_CALL(cudaSetDevice(0));
```
Comment (cp5555) [inline superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_helper.cpp:45]: Only for one device?

File: superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_benchmark.h [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+#include <unordered_map>
+
+#include "cudnn_helper.h"
+
+/**
+ * @brief Enum of cudnn function name
+ */
+namespace cudnn_test {
+enum cudnn_function_name_enum {
+    e_cudnnConvolutionForward = 0,
```
Comment (cp5555) [inline superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_benchmark.h:23]: Why do we only assign 0?

File: superbench/benchmarks/micro_benchmarks/cudnn_test/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,33 @@
+# Copyright (c) Microsoft Corporation.
```
Comment (cp5555) [inline superbench/benchmarks/micro_benchmarks/cudnn_test/CMakeLists.txt:1]: For the folder name, whether do we change it as cublas_function? It's very strange that one is cudnn_test and the other is cublas_function

File: superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_benchmark.h [hunk 5]
```diff
@@ -0,0 +1,254 @@
+// Copyright(c) Microsoft Corporation.
```
Comment (cp5555) [inline superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_benchmark.h:1]: The file name is cudnn_benchmark.h, but it's not related to it. Why not merge this with cudnn_fucntion.h?

File: superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_function_helper.h [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+        function.set_num_test(options.num_test);
+        function.set_warm_up(options.warm_up);
+        function.set_num_in_step(options.num_in_step);
+        function.set_random_seed(options.random_seed);
+        if (function.get_input_type() == CUDNN_DATA_FLOAT && function.get_conv_type() == CUDNN_DATA_FLOAT) {
+            auto p_function = get_cudnn_function_pointer<float, float>(function);
+            p_function->benchmark();
+            delete p_function;
+        }
+        if (function.get_input_type() == CUDNN_DATA_HALF && function.get_conv_type() == CUDNN_DATA_FLOAT) {
```
Comment (cp5555) [inline superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_function_helper.h:187]: else if?

File: superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_function_helper.h [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+            auto p_function = get_cudnn_function_pointer<half, float>(function);
+            p_function->benchmark();
+            delete p_function;
+        }
+        if (function.get_input_type() == CUDNN_DATA_HALF && function.get_conv_type() == CUDNN_DATA_HALF) {
```
Comment (cp5555) [inline superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_function_helper.h:192]: else if?

File: superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_function_helper.h [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+            auto p_function = get_cudnn_function_pointer<half, half>(function);
+            p_function->benchmark();
+            delete p_function;
+        }
```
Comment (cp5555) [inline superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_function_helper.h:196]: Add else for error handling?

File: superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_function_helper.h [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+        random_seed = (random_seed == 0 ? time(NULL) : random_seed);
+        para_info_json = get_cmd_line_argument_string("--config_json");
+        para_info_json =
+            para_info_json == ""
+                ? R"({"algo":0,"arrayLength":2,"conv_type":0,"dilationA":[1,1],"filterStrideA":[1,1],"filter_dims":[32,128,3,3],"input_dims":[32,128,14,14],"input_stride":[25088,196,14,1],"input_type":0,"mode":1, "name":"cudnnConvolutionBackwardFilter","output_dims":[32,32,14,14],"output_stride":[6272,196,14,1],"padA":[1,1],"use_tensor_core":false})"
+                : para_info_json;
+    }
+};
+
+namespace cudnn_test {
```
Comment (cp5555) [inline superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_function_helper.h:100]: Why is Option not in the namespace cudnn_test?

File: superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_function.h [hunk 10]
```diff
@@ -0,0 +1,127 @@
+// Copyright(c) Microsoft Corporation.
```
Comment (cp5555) [inline superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_function.h:1]: Separate class in different file.

File: superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_test.cpp [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+
+ * @param  argc
+ * @param  argv
+ * @return int
+ */
+int main(int argc, char *argv[]) {
+    try {
+        // parse arguments from cmd
+        Options options(argc, argv);
+        // benchmark each function defined in 'para_info.json'
```
Comment (abuccts) [inline superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_test.cpp:38]: remove?

File: superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_test.cpp [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+        cudnn_test::run_benchmark(options);
```
Comment (abuccts) [inline superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_test.cpp:39]: rename to something like "cudnn_kernel_perf"?

File: superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_helper.cpp [hunk 13]
```diff
@@ -0,0 +1,91 @@
+// Copyright (c) Microsoft Corporation.
```
Comment (abuccts) [inline superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_helper.cpp:1]: rename `.cpp` to `.cu`?

File: superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_helper.cpp [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+    // create streams/handles
+    CHECK_CUDNN_ERROR(cudnnCreate(cudnn_handle));
+}
+
+/**
+ * @brief Cuda context free
+ */
+void cuda_free(cudnnHandle_t *cudnn_handle) {
+    CHECK_CUDNN_ERROR(cudnnDestroy(*cudnn_handle));
+    CUDA_SAFE_CALL(cudaSetDevice(0));
```
Comment (abuccts) [inline superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_helper.cpp:55]: why need `cudaSetDevice` in cuda_free?

File: superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_benchmark.h [hunk 15]
```diff
@@ -0,0 +1,254 @@
...
+    }
+    /**
+     * @brief Destroy the Cudnn Function object, including free cuda memory and handle of cudnn and curand
+     */
+    virtual ~CudnnFunction() {
+        // free context and memory
+        CUDA_SAFE_CALL(cudaFree(x));
+        CUDA_SAFE_CALL(cudaFree(filter));
+        CUDA_SAFE_CALL(cudaFree(h));
+        cuda_free(&this->cudnn_handle);
```
Comment (abuccts) [inline superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_benchmark.h:174]: what's the difference between `cudaFree` and `cuda_free`?

File: superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_helper.cpp [hunk 16]
```diff
@@ -0,0 +1,91 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+/**
+ * @file cudnn_helper.cpp
+ * @brief  Cpp file for some functions related to cudnn
+ */
+
+#include <bits/stdc++.h>
+#include <stdlib.h>
```
Comment (abuccts) [inline superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_helper.cpp:10]: are these two libs used in this file?

File: superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_helper.h [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
+        type = CUDNN_DATA_FLOAT;
+    } else if (std::is_same<T, half>::value) {
+        type = CUDNN_DATA_HALF;
+    }
+#if CUDNN_MAJOR >= 6
+    else if (std::is_same<T, uint8_t>::value)
+        type = CUDNN_DATA_INT8;
+#endif
+    else
+        throw("unknown type in tensorDecriptor");
```
Comment (abuccts) [inline superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_helper.h:93]: tensorDecriptor -> tensorDescriptor?

File: superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_benchmark.h [hunk 18]
```diff
@@ -0,0 +1,254 @@
...
+    cudnnDataType_t input_type_;
+    cudnnDataType_t conv_type_;
+    std::string function_str_;
+
+  public:
+    void set_num_test(int num_test) { this->num_test = num_test; }
+    void set_warm_up(int warm_up) { this->warm_up = warm_up; }
+    void set_num_in_step(int num_in_step) { this->num_in_step = num_in_step; }
+    void set_random_seed(int random_seed) { this->random_seed = random_seed; }
+    void set_name(std::string &n) { name = n; }
```
Comment (guoshzhao) [inline superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_benchmark.h:70]: For all reference type, use 'const' type? Like:
`const std::string &n`

File: superbench/benchmarks/micro_benchmarks/cudnn_test/para_info.json (hunk 1 above)
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/cudnn_test/para_info.json]: removed

File: superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_helper.cpp (hunk 2 above)
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_helper.cpp]: I think this code just test for one visible device, you can set cuda_visible_devices to run on other GPU

File: superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_benchmark.h (hunk 3 above)
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_benchmark.h]: it is just a hint

File: superbench/benchmarks/micro_benchmarks/cudnn_test/CMakeLists.txt (hunk 4 above)
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/cudnn_test/CMakeLists.txt]: renamed

File: superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_function_helper.h (hunk 8 above)
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_function_helper.h]: if-else and error handling added

File: superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_function_helper.h (hunk 9 above)
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_function_helper.h]: added option in cudnn_test namespace

File: superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_helper.cpp (hunk 14 above)
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_helper.cpp]: removed

File: superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_benchmark.h (hunk 15 above)
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_benchmark.h]: cudaFree is a api to free the cuda memory of a data pointer, cuda_free is a self-defined function to destroy cudnn handle

File: superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_helper.cpp (hunk 16 above)
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_helper.cpp]: modified to <numeric> and <stdlib.h>, I need to use them for std::rand() and std::accumulate()

File: superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_helper.h (hunk 17 above)
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_helper.h]: yes, thank you for the comment

File: superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_benchmark.h (hunk 18 above)
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_benchmark.h]: ok, thank you for the comment

File: superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_helper.cpp (hunk 2 above)
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_helper.cpp]: I have tested that we can use cuda_visible_devices to decide which gpu to run
image

Comment (abuccts) [reply superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_helper.cpp]: what about `NVIDIA_VISIBLE_DEVICES`?

Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_helper.cpp]: what is NVIDIA_VISIBLE_DEVICES? It does not work.

File: superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_benchmark.h (hunk 3 above)
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_benchmark.h]: removed

File: superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_helper.cpp (hunk 2 above)
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_helper.cpp]: https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/user-guide.html#gpu-enumeration
it works for nvidia docker container I think，not for an application？

File: superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_benchmark.h (hunk 5 above)
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_benchmark.h]: splited them by class

File: superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_test.cpp (hunk 12 above)
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_test.cpp]: TODO: will uniform the name later

File: superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_helper.cpp (hunk 13 above)
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_helper.cpp]: we only use .cu for kernel-related code

File: superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_benchmark.h (hunk 15 above)
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_benchmark.h]: rename to cudnn_handle_free

File: superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_helper.cpp (hunk 16 above)
Comment (yukirora) [reply superbench/benchmarks/micro_benchmarks/cudnn_test/cudnn_helper.cpp]: renamed to <cstdlib>

**Review by cp5555 — APPROVED** [review]

**Review by abuccts — APPROVED** [review]

**Review by guoshzhao — APPROVED** [review]

**Pull request merged**